Tactical/Soldier Control.cpp

```diff
@@ -17805,6 +17741,159 @@ UINT8 SOLDIERTYPE::GetBestEquippedFlashLightRange( )
 	return(bestrange);
 }
 
+bool SOLDIERTYPE::AddBestFlashLight()
+{
+    // not possible to get this bonus on a roof, due to our lighting system
+    if ( this->pathing.bLevel != 0 )
+    {
+        return false;
+    }
+
+    UINT8 maxRange = this->GetBestEquippedFlashLightRange();
+    if ( maxRange < 1 )
+    {
+        return false;
+    }
+
+    // we don't use the flashlight to run better at night (light up our shoes), we use it to find enemies!
+    UINT8 minRange = 4;
+    if ( minRange > maxRange )
+    {
+        minRange = maxRange;
+    }
+
+    float maxAngle = 45;
+    maxAngle *= PI / 180 / 2; // convert to rad and halven
+
+    auto forward = DirectionInc(this->ubDirection);
+    auto left = DirectionInc(DirectionIfTurnedClockwise(this->ubDirection, 6));
+    auto leftLeft = DirectionInc(DirectionIfTurnedClockwise(this->ubDirection, 5));
+    auto right = DirectionInc(DirectionIfTurnedClockwise(this->ubDirection, 2));
+    auto rightRight = DirectionInc(DirectionIfTurnedClockwise(this->ubDirection, 3));
+
+    bool isDiagonal = this->ubDirection == NORTHEAST || this->ubDirection == NORTHWEST || this->ubDirection == SOUTHEAST || this->ubDirection == SOUTHWEST;
+
+	struct position_2d
+	{
+		position_2d(INT32 gridNo)
+		{
+			ConvertGridNoToXY(gridNo, &x, &y);
+		}
+		position_2d(INT16 _x, INT16 _y)
+		{
+			x = _x;
+			y = _y;
+		}
+		INT16 x, y;
+	};
+	struct vector_2d
+	{
+		vector_2d(INT8 direction)
+		{
+			ConvertDirectionToVectorInXY(direction, &dx, &dy);
+			length = sqrt(pow(dx, 2) + pow(dy, 2));
+		}
+		vector_2d(position_2d from, position_2d to)
+		{
+			dx = to.x - from.x;
+			dy = to.y - from.y;
+			length = sqrt(pow(dx, 2) + pow(dy, 2));
+		}

```

Comment:
What should properly do? Cast double to float find a math func using float? I originally wanted to stay max 32bit with computations.

The const & thingy, whatever helps the compiler to optimize. Can i set on the structs to help there? Those are const after setting anyway.

Reply:
the only difference is one extra register load in x86 with double over float. I doubt you could measure the performance difference if you tried. I'd just use double throughout.

you mean the struct members? yea, if they're const, const them. const everything.

Reply:
float vs double: just switched from pow to powf and problem solved. however there are 600+ instances where we use double already. I really though we wanted to keep ja2 32bit optimized, I guess its not :( (in some cases we even use arrays instead of switches to improve performance...)

const does not work as I expected, like with const x, y I can't do `ConvertGridNoToXY(gridNo, &x, &y);` anymore in the constructor.
also found this on the internet:
> size_of(object) < size_of(pointer_to_object)

is that true? can't the compiler do a zero-copy optimization? (e.g. through inlining?)

Reply:
It's behaving as expected: `ConvertGridNoToXY` takes two `INT16 *` arguments, can't pass it it `const INT16`.

If you REALLY want your members to be `const` you could local `INT16 tempX, tempY` to call it, then use those to initialize `const INT16 x,y` in your struct, your call.

Some objects are smaller than the pointer size, yes and I have no idea what the optimizer does 99% of the time.

Reply:
hmm I really wanted to avoid local variables. The whole point for me was to
1. ask another function to directly fill out the members of the struct, allowing for zero-copy performance
2. have those members be const after the constructor ran through, so the compilers knows it can optimize access

I'm too stupid to understand how to do that, meaning I just ignore const for now, until I understand it better

Reply:
Actually I was wrong, you can't assign to const member variables in the constructor body, it has to be in the `ctor() : m{value} {}` initializer list

Reply:
ok keeping as is